File: presentation-manager-web/src/app/model/presentation/presentation.model.ts
import {PanelModel} from '../role/panel.model';
import {SupervisorModel} from '../role/supervisor.model';
import {PresentationMode, ScheduleModel} from '../schedule/schedule.model';
import {TimeRange} from '../../service/test.service';
import {RoomPresentationSlotsModel} from '../room/RoomModel';

export class PresentationModel {
  id: number;
  supervisorId: number;
  studentName: string;
  studentMatrixNo: string;
  studentEmail: string;
  scheduleId: number;
  scheduleTitle: string;
  title: string;
  mark: number;
  resultStatus: string;
  startTime: Date;
  endTime: Date;
  roomId: number;
  roomName: string;
  venue: string;
  mode: PresentationMode;
  scheduleModel: ScheduleModel;
  supervisorModel: SupervisorModel;
  chairperson: PanelModel;
  panelModels: PanelModel[];
  commonAvailabilityList: TimeRange[];

  //google
  calendarId: string;
  calendarHtmlLink: string;
  lastSync: Date;
  googleMeetLink: string;
  calendarStartTime: Date;
  calendarEndTime: Date;

  constructor() {
    this.supervisorModel = new SupervisorModel();
    this.chairperson = new PanelModel();
  }
}

export class PresentationScheduleModel {
  id: number;
  startTime: Date;
  endTime: Date;
  roomId: number;

}

export class SchedulerPresentationModel extends PresentationModel {
  schedulerId: number | string;
  roomId: number;
  scheduleSaved: boolean;
  isBlock: boolean;
}

export class PresentationModeCheckModel extends PresentationModel {
  online: boolean;
  physical: boolean;
}

export class AutoSchedulingModel {
  scheduleId: number;
  presentationsToScheduleOnline: PresentationModel[];
  presentationsToSchedulePhysical: PresentationModel[];
  considerAvailability: boolean;
  roomPresentationSlotsModels: RoomPresentationSlotsModel[];
  onlinePresentationSlotsModel: RoomPresentationSlotsModel;
  presentationDuration: number;
}

export class SchedulerModel {
  presentationToSchedule: PresentationModel[];
  unAvailableTime: PresentationModel[];
}

File: presentationmanager/src/main/java/com/fyp/presentationmanager/model/presentation/PresentationModel.java
package com.fyp.presentationmanager.model.presentation;

import com.fyp.presentationmanager.entity.PresentationBean;
import com.fyp.presentationmanager.entity.PresentationPanelBean;
import com.fyp.presentationmanager.enums.PresentationMode;
import com.fyp.presentationmanager.model.ScheduleModel;
import com.fyp.presentationmanager.model.role.PanelModel;
import com.fyp.presentationmanager.model.role.SupervisorModel;
import com.fyp.presentationmanager.model.scheduleGeneticAlgo.scheduleDNA.TimeRange;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class PresentationModel {
    private Integer id;
    private String studentMatrixNo;
    private String studentName;
    private String studentEmail;
    private Integer scheduleId;
    private String title;
    private Integer mark;
    private Integer resultStatus;
    private Date startTime;
    private Date endTime;

    private Integer roomId;
    private String roomName;
    private PresentationMode mode;
    private ScheduleModel scheduleModel;
    private SupervisorModel supervisorModel;
    private PanelModel chairperson;
    private List<PanelModel> panelModels;
    private List<TimeRange> commonAvailabilityList;
    //google
    private String calendarId;
    private String calendarHtmlLink;
    private Date lastSync;
    private String googleMeetLink;
    private Date calendarStartTime;
    private Date calendarEndTime;

    public static PresentationModel build(PresentationBean presentationBean) {
        PresentationModel presentationModel = new PresentationModel();
        presentationModel.setId(presentationBean.getId());
        presentationModel.setScheduleId(presentationBean.getScheduleId());
        presentationModel.setTitle(presentationBean.getTitle());
        presentationModel.setScheduleId(presentationBean.getScheduleId());
        presentationModel.setScheduleModel(new ScheduleModel(presentationBean.getScheduleBean()));
        presentationModel.setStudentEmail(presentationBean.getStudentEmail());
        presentationModel.setStudentMatrixNo(presentationBean.getStudentMatrixNo());
        presentationModel.setStudentName(presentationBean.getStudentName());
        presentationModel.setSupervisorModel(SupervisorModel.build(presentationBean.getSupervisorBean()));
        if (presentationBean.getChairPersonBean() != null) {
            presentationModel.setChairperson(PanelModel.build(presentationBean.getChairPersonBean()));
        }
        presentationModel.setStartTime(presentationBean.getStartTime());
        presentationModel.setEndTime(presentationBean.getEndTime());
        presentationModel.setRoomId(presentationBean.getRoomId());
        presentationModel.setMode(presentationBean.getMode());
        presentationModel.setCalendarId(presentationBean.getCalendarId());
        presentationModel.setLastSync(presentationBean.getLastSync());
        presentationModel.setGoogleMeetLink(presentationBean.getGoogleMeetLink());
        presentationModel.setCalendarHtmlLink(presentationBean.getCalendarHtmlLink());
        presentationModel.setCalendarStartTime(presentationBean.getCalendarStartTime());
        presentationModel.setCalendarEndTime(presentationBean.getCalendarEndTime());
        if (presentationBean.getRoomBean() != null) {
            presentationModel.setRoomName(presentationBean.getRoomBean().getName());
        }
        List<PanelModel> panelModels = new ArrayList<>();
        if (presentationBean.getPresentationPanelBeans() != null) {
            for (PresentationPanelBean presentationPanelBean : presentationBean.getPresentationPanelBeans()) {
                PanelModel panelModel = PanelModel.build(presentationPanelBean.getPanelBean());
                panelModels.add(panelModel);
            }
        }
        presentationModel.setPanelModels(panelModels);
        return presentationModel;
    }
}
